cluster/service.go

```diff
@@ -491,6 +525,139 @@ func (s *Service) processSeriesKeysRequest(conn net.Conn) {
 	}
 }
 
+func (s *Service) processExpandSourcesRequest(conn net.Conn) {
+	var sources influxql.Sources
+	if err := func() error {
+		// Parse request.
+		var req ExpandSourcesRequest
+		if err := DecodeLV(conn, &req); err != nil {
+			return err
+		}
+
+		// Collect iterator creators for each shard.
+		ics := make([]influxql.IteratorCreator, 0, len(req.ShardIDs))
+		for _, shardID := range req.ShardIDs {
+			ic := s.TSDBStore.ShardIteratorCreator(shardID)
+			if ic == nil {
+				return nil
+			}
+			ics = append(ics, ic)
+		}
+
+		// Expand sources from all shards.
+		a, err := influxql.IteratorCreators(ics).ExpandSources(req.Sources)
+		if err != nil {
+			return err
+		}
+		sources = a
+
+		return nil
+	}(); err != nil {
+		s.Logger.Printf("error reading ExpandSources request: %s", err)
+		EncodeTLV(conn, expandSourcesResponseMessage, &ExpandSourcesResponse{Err: err})
+		return
+	}
+
+	// Encode success response.
+	if err := EncodeTLV(conn, expandSourcesResponseMessage, &ExpandSourcesResponse{
+		Sources: sources,
+	}); err != nil {
+		s.Logger.Printf("error writing ExpandSources response: %s", err)
+		return
+	}
+}
+
+func (s *Service) processBackupShardRequest(conn net.Conn) {
+	if err := func() error {
+		// Parse request.
+		var req BackupShardRequest
+		if err := DecodeLV(conn, &req); err != nil {
+			return err
+		}
+
+		// Backup from local shard to the connection.
+		if err := s.TSDBStore.BackupShard(req.ShardID, req.Since, conn); err != nil {
+			return err
+		}
+
+		return nil
+	}(); err != nil {
+		s.Logger.Printf("error processing BackupShardRequest: %s", err)
+		return
+	}
+}
+
+func (s *Service) processCopyShardRequest(conn net.Conn) {
+	if err := func() error {
+		// Parse request.
+		var req CopyShardRequest
+		if err := DecodeLV(conn, &req); err != nil {
+			return err
+		}
+
+		// Begin streaming backup from remote server.
+		r, err := s.backupRemoteShard(req.Host, req.ShardID, req.Since)
+		if err != nil {
+			return err
+		}
+		defer r.Close()
+
+		// Create shard if it doesn't exist.
+		if err := s.TSDBStore.CreateShard(req.Database, req.Policy, req.ShardID); err != nil {
+			return err
+		}
+
+		// Restore to local shard.
+		if err := s.TSDBStore.RestoreShard(req.ShardID, r); err != nil {
+			return err
+		}
+
+		return nil
+	}(); err != nil {
+		s.Logger.Printf("error reading CopyShard request: %s", err)
+		EncodeTLV(conn, copyShardResponseMessage, &CopyShardResponse{Err: err})
+		return
+	}
+
+	// Encode success response.
+	if err := EncodeTLV(conn, copyShardResponseMessage, &CopyShardResponse{}); err != nil {
+		s.Logger.Printf("error writing CopyShard response: %s", err)
+		return
```

Comment:
I defensively added that in case code was added in the future after the `if` block. That's just my personal preference though.